### PR TITLE
Expose Rancher authentication as a k8s authentication webhook endpoint

### PR DESCRIPTION
**Pull request by erikwilson**
For issue: rancher/rancher/issues/13698

Related authentication service: https://github.com/erikwilson/rancher-kube-api-auth/tree/rke-remote-kubectl

TODO:
- [X] Refactor api-auth service to golang standards
- [X] Better vendoring in api-auth service
- [X] Clean up Kubeconfig generation
- [X] Cluster scoped tokens using rancher token types
- [X] PRs for related dependencies
- [X] Tests

### CODE REVIEW

File: vendor/github.com/rancher/rke/cluster/authn-webhook.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+			"sh",
+			"-c",
+			fmt.Sprintf("t=$(mktemp); echo -e \"%s\" > $t && mv $t %s && chmod 644 %s", yaml, AuthnWebhookConfigFile, AuthnWebhookConfigFile),
+		},
+	}
+	hostCfg := &container.HostConfig{
+		Binds: []string{
+			fmt.Sprintf("%s:/etc/kubernetes:z", path.Join(host.PrefixPath, "/etc/kubernetes")),
+		},
+		Privileged: true,
```
Comment (cjellick) [inline vendor/github.com/rancher/rke/cluster/authn-webhook.go]: why does it need to be privileged?

Comment (erikwilson) [reply vendor/github.com/rancher/rke/cluster/authn-webhook.go]: Fixed, it shouldn't need to be, probably same for cloud-provider.go

File: pkg/clustermanager/manager.go [hunk 2]
```diff
@@ -173,6 +174,13 @@ func (m *Manager) doStart(rec *record, clusterOwner bool) (exit error) {
 		}
 	}()
 
+	if rec.clusterRec.Spec.EnableClusterAuth {
```
Comment (cjellick) [inline pkg/clustermanager/manager.go]: move to where we register other controlelrs

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go [hunk 3]
```diff
@@ -109,11 +132,11 @@ func (cd *clusterDeploy) deployAgent(cluster *v3.Cluster) error {
 	}
 
 	_, err = v3.ClusterConditionAgentDeployed.Do(cluster, func() (runtime.Object, error) {
-		yaml, err := cd.getYAML(cluster, desired)
+		yaml, err := cd.getYAML(cluster, desiredAgent, desiredAuth)
 		if err != nil {
 			return cluster, err
 		}
-
+		logrus.Info("YAML:\n", string(yaml))
```
Comment (cjellick) [inline pkg/controllers/management/clusterdeploy/clusterdeploy.go]: remove this log statement

File: pkg/controllers/management/clusterdeploy/clusterdeploy.go [hunk 4]
```diff
@@ -172,7 +207,10 @@ func (cd *clusterDeploy) getKubeConfig(cluster *v3.Cluster) (*clientcmdapi.Confi
 	return cd.clusterManager.KubeConfig(cluster.Name, token), nil
 }
 
-func (cd *clusterDeploy) getYAML(cluster *v3.Cluster, agentImage string) ([]byte, error) {
+func (cd *clusterDeploy) getYAML(cluster *v3.Cluster, agentImage, authImage string) ([]byte, error) {
+	logrus.Info("Desired agent image:", agentImage)
```
Comment (cjellick) [inline pkg/controllers/management/clusterdeploy/clusterdeploy.go]: probably should be debug level?

**Comment by cjellick** [comment]
Can you rebase and rework your commits to make this easier to digest?
I think some of the rke vendor changes that are showing up in your PR are actually from unrelated RKE updates.
I'm mainly looking for:
1. rebase
2. separate vendor updates from non-vendor code commits
3. separate script/packaging changes from code changes
4. squash into logical commits, instead of  a bunch of WIP commits. You dont need to get overly granular. If most of the cod changes are interdependent, a single commit (excluding the things mentioned above) is fine.

File: scripts/build-server [hunk 5]
```diff
@@ -8,6 +8,6 @@ cd $(dirname $0)/..
 mkdir -p bin
 [ "$(uname)" != "Darwin" ] && LINKFLAGS="-extldflags -static -s"
 
-RKE_VERSION=$(grep '^github.com/rancher/rke' vendor.conf | awk '{print $2}')
+RKE_VERSION=$(grep -m1 '^github.com/rancher/rke' vendor.conf | awk '{print $2}')
```
Comment (dramich) [inline scripts/build-server]: Why was this added? We should only have 1 version in vendor.conf

File: scripts/test [hunk 6]
```diff
@@ -1,41 +1,37 @@
...
 set -e
 
+CWD=$(pwd -P)
 cleanup()
 {
     EXIT=$?
-    kill $PID
+    set +e
+    echo Stopping rancher server
+    kill %%
```
Comment (dramich) [inline scripts/test]: This should be explicit, we can't guarantee it's rancher that is going to get the axe here

File: scripts/test [hunk 7]
```diff
@@ -1,41 +1,37 @@
...
+    wait
```
Comment (dramich) [inline scripts/test]: Use the PID that's going to feed `kill` here

File: scripts/build-server (hunk 5 above)
Comment (erikwilson) [reply scripts/build-server]: For development & testing purposes it is helpful to have more than one, where a dev version may be defined at the top of vendor.conf. As trash selects the first version, placing the dev version at the top will avoid conflicts if rebasing.

File: tests/core/cluster_common.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    - controlplane
+    - worker
+    - etcd{% endfor %}
+"""
+    return dind_cluster_config_j2
+
+
+def create_cluster(cluster_config_file):
+    # attempt to resolve unknown random rke up errors
+    for _ in range(3):
```
[the remark this answers is not in the document]
Comment (erikwilson) [reply tests/core/cluster_common.py]: Not sure why RKE up sometimes fails, it looks like a client timeout issue. @galal-hussein might have some input, with the modified scripts we search for appropriate config files to clean up any instances that the python code did not remove. With the loop using the same config file I don't think there is a possibility to create multiple instances in the loop.

File: tests/core/cluster_common.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            return subprocess.check_output(
+                'rke up --dind --config ' +
+                cluster_config_file, shell=True
+            )
+        except subprocess.CalledProcessError as err:
+            print('RKE up error: ' + str(err.returncode))
+    raise Exception('rke up failure')
+
+
+def remove_cluster(cluster_config_file):
```
[the remark this answers is not in the document]
Comment (erikwilson) [reply tests/core/cluster_common.py]: I haven't seen RKE down fail

File: scripts/test (hunk 6 above)
Comment (erikwilson) [reply scripts/test]: Updated

File: scripts/test (hunk 7 above)
Comment (erikwilson) [reply scripts/test]: Updated

File: tests/core/conftest.py [hunk 10]
```diff
@@ -151,6 +193,42 @@ def admin_pc_client(admin_pc):
     return admin_pc.client
 
 
+@pytest.fixture(scope="session")
+def dind_cc(request, admin_mc):
+    # make sure server-url is set
```
Comment (dramich) [inline tests/core/conftest.py]: Check what os you are running on and warn the user if it's not going to work

File: tests/core/test_cluster_auth_tokens.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        try:
+            exec_kubectl(dind_cc, headers)
+            return True
+        except Exception:
+            return False
+    return test
+
+
+# as an admin, we should have access
+def test_admin_api_resources(dind_cc):
```
Comment (dramich) [inline tests/core/test_cluster_auth_tokens.py]: Mark the tests to not run if we are not in an OS we support with DnD, mark, skip-if through py.test etc.

File: tests/core/cluster_common.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+import subprocess
+import os
+import re
+import yaml
+from jinja2 import Template
+
+
+def generate_cluster_config(dind_rke_node_num):
+    # generate a random and kube_config file
+    dind_random = random.randint(1, 100000)
```
Comment (dramich) [inline tests/core/cluster_common.py]: Use the func from common `random_num`

File: tests/core/cluster_common.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    dind_name = 'dind-' + str(dind_random)
+    dind_cluster_config_file = dind_name + '.yml'
+    dind_kube_config_file = 'kube_config_' + dind_name + '.yml'
+    cluster_config_tmpl = get_rke_config_template()
+    # generate nodes
+    random_nodes = [
+            'node-' +
+            str(random.randint(10000, 99999))
```
Comment (dramich) [inline tests/core/cluster_common.py]: use the func from common `random_str`

File: tests/core/cluster_common.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    - etcd{% endfor %}
+"""
+    return dind_cluster_config_j2
+
+
+def create_cluster(cluster_config_file):
+    # attempt to resolve unknown random rke up errors
+    for _ in range(3):
+        try:
+            return subprocess.check_output(
```
Comment (dramich) [inline tests/core/cluster_common.py]: Catch your error output by adding `stderr=subprocess.STDOUT` to the call to `check_output` then your error message will be there.

File: tests/core/cluster_common.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                cluster_config_file, shell=True
+            )
+        except subprocess.CalledProcessError as err:
+            print('RKE up error: ' + str(err.returncode))
+    raise Exception('rke up failure')
+
+
+def remove_cluster(cluster_config_file):
+    try:
+        return subprocess.check_output(
```
Comment (dramich) [inline tests/core/cluster_common.py]: Catch your error output by adding `stderr=subprocess.STDOUT` to the call to `check_output` then your error message will be there.

File: tests/core/cluster_common.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+def create_cluster(cluster_config_file):
+    # attempt to resolve unknown random rke up errors
+    for _ in range(3):
+        try:
+            return subprocess.check_output(
+                'rke up --dind --config ' +
+                cluster_config_file, shell=True
+            )
+        except subprocess.CalledProcessError as err:
+            print('RKE up error: ' + str(err.returncode))
```
Comment (dramich) [inline tests/core/cluster_common.py]: Update to print the error

File: tests/core/cluster_common.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def remove_cluster(cluster_config_file):
+    try:
+        return subprocess.check_output(
+            'rke remove --force --dind --config ' +
+            cluster_config_file, shell=True
+        )
+    except subprocess.CalledProcessError as err:
+        print('RKE down error: ' + str(err.returncode))
```
Comment (dramich) [inline tests/core/cluster_common.py]: Update to print the error

File: tests/core/conftest.py [hunk 18]
```diff
@@ -269,3 +347,55 @@ def protect_response(r):
...
+    admin_mc.client.delete(cluster)
+    remove_cluster(cluster_file)
+
+
+def create_kubeconfig(dind_cc, headers):
+    # wait for permissions to sync
+    time.sleep(1)
+
+    # request cluster scoped kubeconfig
+    KUBE_URL = BASE_URL + \
```
Comment (dramich) [inline tests/core/conftest.py]: This should use the client, you can get actions off `dind_cc.cluster.actions.<action_name>`

File: tests/core/conftest.py [hunk 19]
```diff
@@ -269,3 +347,55 @@ def protect_response(r):
...
+        '/clusters/' + \
+        dind_cc.cluster.id + \
+        '?action=generateKubeconfig'
+    r = requests.post(KUBE_URL, headers=headers, verify=False)
+    protect_response(r)
+    cluster_kubeconfig = r.json()['config']
+
+    # write cluster scoped kubeconfig
+    cluster_kubeconfig_file = "kubeconfig-" + \
```
Comment (dramich) [inline tests/core/conftest.py]: Use a finalizer to cleanup the files so they aren't left around

File: tests/core/conftest.py [hunk 20]
```diff
@@ -269,3 +347,55 @@ def protect_response(r):
...
+    def cluster_token_available():
+        try:
+            return crd_client.get_namespaced_custom_object(
+                'cluster.cattle.io',
+                'v3',
+                'cattle-system',
+                'clusterauthtokens',
+                token_name
+            )
+        except Exception:
```
Comment (dramich) [inline tests/core/conftest.py]: This should catch `ApiException`, naked `Exception` will mask anything else that could throw.

File: tests/core/test_cluster_auth_tokens.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+# test if the kubeconfig works to list api-resources for the fqdn context
+def exec_kubectl(dind_cc, headers, cmd='api-resources'):
+    cluster_kubeconfig_file = create_kubeconfig(dind_cc, headers)
+    # verify cluster scoped access
+    kubectl_cmd = 'kubectl ' + cmd + \
+                  ' --kubeconfig ' + cluster_kubeconfig_file + \
+                  ' --context ' + dind_cc.name + '-fqdn'
+    try:
+        return subprocess.check_output(kubectl_cmd, shell=True)
```
Comment (dramich) [inline tests/core/test_cluster_auth_tokens.py]: Catch your error output by adding stderr=subprocess.STDOUT and update the print statement

File: tests/core/test_cluster_auth_tokens.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        raise err
+
+
+# test generator for multiple attempts
+def kubectl_available(dind_cc, headers):
+    def test():
+        try:
+            exec_kubectl(dind_cc, headers)
+            return True
+        except Exception:
```
Comment (dramich) [inline tests/core/test_cluster_auth_tokens.py]: Specify the error type you are expecting here so other exceptions are not masked.

File: tests/core/test_cluster_auth_tokens.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    wait_for(kubectl_available(dind_cc, dind_cc.admin_mc.client._headers))
+
+
+# as a user which has not been given permission, we should fail
+def test_user_no_template(dind_cc, user_mc):
+    test_admin_api_resources(dind_cc)
+    try:
+        exec_kubectl(dind_cc, user_mc.client._headers)
+        assert False
+    except Exception:
```
Comment (dramich) [inline tests/core/test_cluster_auth_tokens.py]: This should be a specific exception. What error is expected here? Does the user still get the 2nd context with no permissions? If not then it's probably a `kubectl` error for the context not existing.

File: tests/core/test_cluster_auth_tokens.py [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+def test_admin_api_resources(dind_cc):
+    wait_for(kubectl_available(dind_cc, dind_cc.admin_mc.client._headers))
+
+
+# as a user which has not been given permission, we should fail
+def test_user_no_template(dind_cc, user_mc):
+    test_admin_api_resources(dind_cc)
+    try:
+        exec_kubectl(dind_cc, user_mc.client._headers)
+        assert False
```
Comment (dramich) [inline tests/core/test_cluster_auth_tokens.py]: Add a comment as to why this fails

File: tests/core/conftest.py [hunk 25]
```diff
@@ -269,3 +347,55 @@ def protect_response(r):
...
+
+
+def cleanup_dind(admin_mc, cluster, cluster_file):
+    admin_mc.client.delete(cluster)
+    remove_cluster(cluster_file)
+
+
+def create_kubeconfig(dind_cc, headers):
+    # wait for permissions to sync
+    time.sleep(1)
```
Comment (dramich) [inline tests/core/conftest.py]: If a wait is required use a loop to check on things

File: pkg/controllers/user/clusterauthtoken/setting.go [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	managementv3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const settingName = "auth-provider-refresh-debounce-seconds"
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/setting.go]: This could be moved to common and exported then rancher-kube-api-auth can take advantage of it.

File: pkg/controllers/user/clusterauthtoken/setting.go [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+	return nil, err
+}
+
+func (h *SettingHandler) Updated(setting *managementv3.Setting) (runtime.Object, error) {
+	if setting.Name != settingName {
+		return nil, nil
+	}
+
+	config, err := h.clusterConfigMapLister.Get(h.namespace, setting.Name)
+	if errors.IsNotFound(err) {
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/setting.go]: Do the `err != nil` check first, in that block check the error type

```
if err != nil {
    if !errors.IsNotFound(err) {
        return nil, err
    }
    return h.Create(setting)
}
````

File: pkg/controllers/user/clusterauthtoken/setting.go [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+		return h.Create(setting)
+	}
+	if config == nil || err != nil {
+		return nil, err
+	}
+	if config.Data["value"] == setting.Value {
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/setting.go]: This has potential to panic, check it exists, then the value.

File: pkg/controllers/user/clusterauthtoken/token.go [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+	for _, gp := range userAttribute.GroupPrincipals {
+		for i := range gp.Items {
+			groups = append(groups, gp.Items[i].Name)
+		}
+	}
+	sort.Strings(groups)
+
+	userEnabled := user.Enabled == nil || *user.Enabled
+
+	clusterUserAttribute, err := h.clusterUserAttributeLister.Get(h.namespace, userID)
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/token.go]: Move the error check here. If it doesn't exist create it, if it does, do the update check.

File: pkg/controllers/user/clusterauthtoken/user.go [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func (h *UserHandler) Create(user *managementv3.User) (runtime.Object, error) {
+	return nil, nil
+}
+
+func (h *UserHandler) Updated(user *managementv3.User) (runtime.Object, error) {
+	clusterUserAttribute, err := h.clusterUserAttributeLister.Get(h.namespace, user.Name)
+	if errors.IsNotFound(err) {
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/user.go]:
```
if err != nil {
    if !errors.IsNotFound(err) {
        return nil, err
    }
    return nil, nil
}
````

File: pkg/controllers/user/clusterauthtoken/user.go [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+		return nil, nil
+	}
+
+	clusterUserAttribute.Enabled = userEnabled
+	_, err = h.clusterUserAttribute.Update(clusterUserAttribute)
+	return nil, err
+}
+
+func (h *UserHandler) Remove(user *managementv3.User) (runtime.Object, error) {
+	clusterUserAttribute, err := h.clusterUserAttributeLister.Get(h.namespace, user.Name)
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/user.go]: Just call delete and check the error

File: pkg/controllers/user/clusterauthtoken/user_attribute.go [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+	enabled      bool
+}
+
+func (h *UserAttributeHandler) Create(userAttribute *managementv3.UserAttribute) (runtime.Object, error) {
+	return nil, nil
+}
+
+func (h *UserAttributeHandler) Updated(userAttribute *managementv3.UserAttribute) (runtime.Object, error) {
+	clusterUserAttribute, err := h.clusterUserAttributeLister.Get(h.namespace, userAttribute.Name)
+	if errors.IsNotFound(err) {
```
Comment (dramich) [inline pkg/controllers/user/clusterauthtoken/user_attribute.go]:
```
if err != nil {
    if !errors.IsNotFound(err) {
        return nil, err
    }
...
````

File: pkg/controllers/user/controllers.go [hunk 33]
```diff
@@ -34,17 +35,17 @@ import (
...
 	"github.com/rancher/types/factory"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	// init upgrade implement
 	_ "github.com/rancher/rancher/pkg/controllers/user/logging/upgrade"
 	_ "github.com/rancher/rancher/pkg/controllers/user/pipeline/upgrade"
 )
 
-func Register(ctx context.Context, cluster *config.UserContext, kubeConfigGetter common.KubeConfigGetter, clusterManager healthsyncer.ClusterControllerLifecycle) error {
+func Register(ctx context.Context, cluster *config.UserContext, clusterRec *managementv3.Cluster, kubeConfigGetter common.KubeConfigGetter, clusterManager healthsyncer.ClusterControllerLifecycle) error {
```
Comment (dramich) [inline pkg/controllers/user/controllers.go]: This looks like a good change.

File: tests/core/test_cluster_auth_tokens.py [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+def test_admin_api_resources(request, dind_cc):
+    wait_for(kubectl_available(request, dind_cc, dind_cc.admin_mc.client))
+
+
+# as a user which has not been given permission, we should fail
+@pytest.mark.skipif(platform != 'linux', reason='requires linux for dind')
+@pytest.mark.nonparallel
+def test_user_no_template(request, dind_cc, user_mc):
+    test_admin_api_resources(request, dind_cc)
+    try:
```
Comment (cjellick) [inline tests/core/test_cluster_auth_tokens.py]: a better pattern is this:
```
    with pytest.raises(ApiError) as e:
                exec_kubectl(request, dind_cc, user_mc.client)
    assert e.value.error.status == 401 (or whatever)
```

**Comment by cjellick** [comment]
LGTM
